runtime/io.h

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Defines API between compiled code and I/O runtime library.
+
+#ifndef FORTRAN_RUNTIME_IO_H_
+#define FORTRAN_RUNTIME_IO_H_
+
+#include "entry-names.h"
+#include <cinttypes>
+#include <cstddef>
+
+namespace Fortran::runtime {
+class Descriptor;
+class NamelistGroup;
+};
+
+namespace Fortran::runtime::IO {
+
+class IOStatementState;
+using Cookie = IOStatementState *;
+using ExternalUnit = int;
+using AsynchronousId = int;
+static constexpr ExternalUnit DefaultUnit{-1};  // READ(*), WRITE(*), PRINT
+
+extern "C" {
+
+#define IONAME(name) RTNAME(IO_##name)
+
+// These functions initiate data transfer statements (READ, WRITE, PRINT).
+// Example: PRINT *, 666 is implemented as the series of calls:
+//   Cookie cookie{BeginExternalListOutput(DefaultUnit)};
+//   OutputInteger64(cookie, 666);
+//   EndIOStatement(cookie);
+
+// Internal I/O initiation
+// Internal I/O can loan the runtime library an optional block of memory
+// in which the library can maintain state across the calls that implement
+// the internal transfer; use of these blocks can reduce the need for dynamic
+// memory allocation &/or thread-local storage.  The block must be sufficiently
+// aligned to hold a pointer.
+constexpr std::size_t RecommendedInternalIOScratchAreaBytes(
+    int maxFormatParenthesesNestingDepth) {
+  return 32 + 8 * maxFormatParenthesesNestingDepth;
```

Comment:
It might be possible to make this largely foolproof by defining a call to query the library for this value.  That could effectively guarantee that this storage can be used 100% of the time.  The question is when to make the call, and where to put the value.  One possibility would be to make the call in the prolog of any subprogram that has one or more internal I/O calls and store it local to each such routine.  That would avoid introducing even a weak link between compiler and library versions.

Reply:
We don't want compiled code to have to allocate & deallocate heap space for this, and `alloca()` is not available in all environments, so this size really needs to be known at compilation time, I think.